sound/soc/sof/intel/hda-pcm.c

```diff
@@ -95,6 +95,83 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
 	}
 };
 
+/* There are two dma modes in dsp:
+ * (1) host dma and link dma work in decouple mode for SOF + HDA
+ * (2) host dma and GP dma work in couple mode for SOF + I2S
+ * The difference is that in decopule mode, the host dma is set by dai
+ * frontend and link dma is set by dai backend and the setting maybe
+ * different while in couple mode, host dma and GP dma share the same
+ * dma setting.
+ */
+static inline int get_be_dma_channel(struct snd_sof_dev *sdev,
+				     struct snd_pcm_substream *substream,
+				     struct sof_ipc_stream_params *ipc_params)
+{
+	struct snd_soc_pcm_runtime *fe = substream->private_data;
+	struct hdac_ext_stream *be_stream = NULL;
+	struct snd_soc_dpcm *dpcm;
+	int direction = substream->stream;
+	struct sof_intel_be_params *be_params;
+
+	be_params = (struct sof_intel_be_params *)ipc_params->private_data;
+	be_params->be_count = 0;
+
+	/* travel BE clients to get BE stream with link dma id */
+	list_for_each_entry(dpcm, &fe->dpcm[direction].be_clients, list_be) {
+		struct snd_soc_pcm_runtime *be = dpcm->be;
+		struct snd_soc_dapm_widget *dapm_widget;
+		struct snd_sof_widget *sof_widget;
+		int index = be_params->be_count;
+
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+			dapm_widget = be->cpu_dai->playback_widget;
+		else
+			dapm_widget = be->cpu_dai->capture_widget;
+
+		sof_widget = snd_sof_find_swidget(sdev, dapm_widget->name);
+		if (!sof_widget) {
+			dev_err(sdev->dev, "error: failed to find backend widget");
+			return -EINVAL;
+		}
+
+		be_stream = snd_soc_dai_get_dma_data(be->cpu_dai, substream);
+		if (be_stream) {
+			/* link dma channel is derived from stream_tag
+			 * in decouple mode
+			 */
+			be_params->be_dma_ch[index] =
+					be_stream->hstream.stream_tag - 1;
+		} else {
+			/* the setting of GP dma is the same as host dma
+			 * in default. Actually, this value is ignored by
+			 * allocation function for GP dma
+			 */
```

Comment:
Again this makes no sense, the settings of GP DMA and host DMA are completely independent. The former are not known to the driver.
Also why do you need to set a value if it's ignored?

Reply:
yes, I set it just to initialize a default value. It may safe to reduce some bugs.
I will remove it.